hash_practice/exercises.py

```diff
@@ -2,18 +2,84 @@
 def grouped_anagrams(strings):
     """ This method will return an array of arrays.
         Each subarray will have strings which are anagrams of each other
-        Time Complexity: ?
-        Space Complexity: ?
+        Time Complexity: On^4 (?)
+        Space Complexity: On
     """
-    pass
+    groupings = []
+
+    j = 0
+    while j < len(strings):
+
+        letter_map = {}
+        sub_group = [strings[j]]
+
+        for letter in strings[j]:
+            if letter not in letter_map:
+                letter_map[letter] = 1
+            else:
+                letter_map[letter] += 1
+
+        k = j + 1
+        while k < len(strings):
+
+            comparison_letter_map = {}
+
+            for letter in strings[k]:
+                if letter not in comparison_letter_map:
+                    comparison_letter_map[letter] = 1
+                else:
+                    comparison_letter_map[letter] += 1
+
+            if comparison_letter_map == letter_map:
+                sub_group.append(strings[k])
+                strings.remove(strings[k])
+                k = k
+            else:
+                k += 1
+        
+        groupings.append(sub_group)
+
+        j += 1
+
+    return groupings
+
 
 def top_k_frequent_elements(nums, k):
     """ This method will return the k most common elements
         In the case of a tie it will select the first occuring element.
-        Time Complexity: ?
-        Space Complexity: ?
+        Time Complexity: On^2
+        Space Complexity: On
     """

```

Comment:
👍 This works, but the runtime is O(nk + n log n), due to the sort and the final set of nested loops.  I think you could do this in O(nk) time or O(n log n).

You could either sort and select the last k elements, or make a loop going through which iterates k times and selects and removes the max frequency each time.